trait2d/analysis/__init__.py

```diff
@@ -1049,26 +1054,35 @@ def normalized(self, normalize_t = True, normalize_xy = True):
         return NormalizedTrack(x, y, t, xmin, ymin, tmin, id=self._id)
 
     def calculate_msd(self):
-        """Calculates the mean squared displacement of the track. Result is stored in the Track object.
+        
+        """Calculates the track's mean squared displacement (msd) and stores it in 'track' object. Also deletes temporary values and colelcts them.
+           Furthermore, calculates the standard deviation per point of the MSD array (msd_error) and the standard error of the mean (SEM) [legacy version].
         """
+        
+        import gc
+                
+        N = len(self._x)
+        col_Array  = np.zeros(N-3)
+        Err_col_Array = np.zeros(N-3)
+
+        data_tmp = np.column_stack((self._x, self._y))
+    
+        for i in tqdm.tqdm(range(1,N-2)):
+            calc_tmp = np.sum(np.abs((data_tmp[1+i:N,:] - data_tmp[1:N - i,:]) ** 2), axis=1)
+            col_Array[i-1] = np.mean(calc_tmp)
+            Err_col_Array[i-1] = np.std(calc_tmp)
+            
+            del calc_tmp
 
-        def circular_matrix(ar):
-            a = np.concatenate(( ar, ar[:-1] ))
-            L = len(ar)
-            n = a.strides[0]
-            return np.lib.stride_tricks.as_strided(a[L-1:], (L,L), (-n,n))
-
-        xc = circular_matrix(self._x)
-        yc = circular_matrix(self._y)
-
-        sdm = (self._x - xc)**2 + (self._y - yc)**2
-        sdm = np.triu(sdm, 1)[:-3]
+        d = np.arange(self._x.shape[0] - 1, 2, -1)
 
-        # MSD is the mean of each upper triangular row (except the last)
-        d = np.arange(sdm.shape[1] - 1, 2, -1)
+        self._msd = col_Array                       #store MSD
+        self._msd_error = Err_col_Array             #store stdev of MSD
+        self._msd_SEM = Err_col_Array/(np.sqrt(d))  #calculate and store SEM of MSD
+        
+        del col_Array, Err_col_Array, data_tmp, N, d
```

Comment:
These are all local variables and will be deleted automatically when the function is exited. No need to delete them manually.

I think there is almost never a reason to use `del` on local variables. Variables in Python are just references to the underlying data (in contrast to C, for example). Deleting the variable does not delete the data but only the reference to it. Variables are automatically deleted when their scope is exited and Python automatically deletes objects/data when no reference to it exists anymore. This is called reference counting.

Calling `del` is pretty much the same as assigning `None` (it simply has the additional effect of preventing you from referencing the variable name again without re-declaring it).

Reply:
You are right here, dear Sir.